---
title: "Computational single-cell biology course"
author: "Hakime Öztürk (h.oeztuerk@dkfz-heidelberg.de)"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    BiocStyle::html_document:    
     code_download: true
     toc: yes

---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

colorize <- function(x, color) {
  if (knitr::is_latex_output()) {
    sprintf("\\textcolor{%s}{%s}", color, x)
  } else if (knitr::is_html_output()) {
    sprintf("<span style='color: %s;'>%s</span>", color, 
      x)
  } else x
}
```
# Dimensionality reduction and clustering on scRNA-seq data

We will work on scRNA-seq data of mouse gastrulation and early organogenesis from [Pijuan-Sala et al., 2019](https://www.nature.com/articles/s41586-019-0933-9). [This Shiny application](https://marionilab.cruk.cam.ac.uk/MouseGastrulation2018/)  provides an interactive interface that allows users to validate their own analysis on this data. You can reach the original data and related scripts from the [Github page](https://github.com/MarioniLab/EmbryoTimecourse2018).

> Gastrulation is a phase early in the embryonic development of most animals, during which the single-layered blastula is reorganized into a multilayered structure known as the gastrula [(Wikipedia, 2020-06-02)](https://en.wikipedia.org/wiki/Gastrulation). 


## Getting familiar with the pre-processed data

Let's start with including required libraries. 
```{r libraries}
  suppressPackageStartupMessages({
  library(Seurat)
  library(ggplot2)
  library(dplyr)
  library(data.table)
  library(cowplot)
  set.seed(1)
})
```

```{r convert, include=FALSE}
 # /Applications/RStudio.app/Contents/MacOS/RStudio
  # here I am reading a SCE object and convert it to Seurat object
  #mgscsce <- readRDS('data/gastrulation/SingleCellExperiment.rds')
  #mgsc <- as.Seurat(mgscsce, counts = "counts", data = "logcounts", project = "mouse gastrulation") 
  
  # I am saving this as Seurat object
  #saveRDS(mgsc, file = "data/gastrulation/mgsc.rds")

```

In the previous practical, we have learned the essential preprocessing steps for working with scRNA-seq. Here we will start working with the preprocessed version of the mouse gastrulation scRNA-seq data. We will load the pre-saved `Seurat` object of this data. 

**Warning:** The pre-processed mouse gastrulation scRNA-seq Seurat object is large (\tilde 2GBs). In today's practical, we will subset the data for fast computation. You can, however, work on the [original data](ftp://ftp.dkfz-heidelberg.de/outgoing/course_material_day_1/mgsc.rds) if you want to practise what we learned today. The [meta-data](ftp://ftp.dkfz-heidelberg.de/outgoing/course_material_day_1/sample_metadata.txt.gz) is provided seperately. 

```{r read}
    mgsc <- readRDS('data/gastrulation/mgsc.rds')
    mgsc
```   
  **Reminder:** number of rows = genes = features, and number of columns = cells = samples

```{r slot}    
    # let's see the available slots
    slotNames(mgsc)
```

Now let's look at to the metadata table of this dataset that contains an overview of the samples. 

```{r metadataadd}
    # see it is empty for now
    head(mgsc@meta.data, 5)

    # now let's load the metadata 
    metadata <- fread('data/gastrulation/sample_metadata.txt.gz') %>% .[stripped==FALSE & doublet==FALSE]
    
    # and add  the metadata  to our seurat object
    mgsc <- AddMetaData(mgsc, metadata = data.frame(metadata, row.names = metadata$cell))
    
    # now let's see once more
    head(mgsc@meta.data, 5)
    
```

Now we are able to see the annotations (e.g. cell type) for each cell. There is also a column named `stage` which shows the embryonic day the cells were sequenced. To speed up our experiments, we will work on a subset of cells that belong to stage `E6.75`. 

```{r subset}
    mgsc_subset <- mgsc[ , mgsc@meta.data$stage=='E6.75']
    mgsc_subset
    # now lets save this subset
    saveRDS(mgsc_subset, file = "data/gastrulation/mgsc_e675.rds")
    
    mgsc <- mgsc_subset
  
```

This is where you are starting from! :) Now you can load [`mgsc_e675.rds`](ftp://ftp.dkfz-heidelberg.de/outgoing/course_material_day_1/mgsc_e675.rds)  object to start working! Make sure that the object you loaded has same numbers of columns and rows with `E6.75` Seurat object.

**Q1: Let's warm up! Can you try to subset cells in locations  $1, 23, 515$ and genes in locations $21, 44, 116$, respectively?** 

<details>
<summary>**Answer**</summary>
```{r p1}
    s1 <- mgsc[, c(1,23,515)] # Subset to cells 1, 23, 515
    s2 <- mgsc[c(21,44,116), ] # Subset to genes 21, 44, 116 
    
    s1
    s2
```
</details>


We can use `rownames` and `colnames` to see the names of genes and cells. 

```{r p2}
# rownames = gene names
head(rownames(mgsc))

# colnames = sample/cell names
head(colnames(mgsc))
```

**Q2: Observe the count data of the first 15 cells for genes "ENSMUSG00000051951" and "ENSMUSG00000033845". ** (` Hint:` use `GetAssayData` function. )

<details>
<summary>**Answer**</summary>
```{r p3}
GetAssayData(object = mgsc)[c("ENSMUSG00000051951", "ENSMUSG00000033845"), 1:15]
# mgsc[c("ENSMUSG00000051951", "ENSMUSG00000033845"), 1:15]@assays$RNA@counts
```
</details>

The . values in the matrix represent $0$s. Since most values in an scRNA-seq matrix are $0$, Seurat uses a sparse-matrix representation to speed up calculations and reduce memory usage.


#  Dimensionality reduction

## Feature selection: identification of highly variable features

**Q3:Our feature set (i.e. number of genes) are quite large. What kind of features do you think should be in the dataset?** 

<details>
<summary>**Answer**</summary>
`r colorize("We will choose a subset of features that have high cell-to-cell variation in the dataset (i.e, genes that are highly expressed in some cells, and lowly expressed in others) which can help highlighting biological signal.", "red")` 
</details>

Seurat implements `FindVariableFeatures` to model the mean-variance relationship in single-cell data which reduces to dimensions to 2000 by default. We will update `nfeatures` argument to reduce dimensionality to 1000. 

```{r findvariable}
mgsc <- FindVariableFeatures(mgsc, selection.method = "vst", nfeatures = 1000)
options(repr.plot.width=12, repr.plot.height=6)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(mgsc), 10)

# plot variable features 
plot1 <- VariableFeaturePlot(mgsc)
LabelPoints(plot = plot1, points = top10, repel = TRUE,  xnudge = 0, ynudge = 0)

head(HVFInfo(mgsc)[VariableFeatures(mgsc), ], 5)
```

### Scaling the data 

**Q4: Seurat implements the `ScaleData` function to scale the data.  Do you remember why we need this step?** 

```{r scale}
mgsc <- ScaleData(mgsc) #Vector of features names to scale/center. Default is variable features.
```

The results are stored in` mgsc[["RNA"]]@scale.data`.

```{r scalehead} 
mgsc[["RNA"]]@scale.data[1:5,1:5]
# OR
# mgsc@assays$RNA@scale.data[1:5,1:5]

```


## Principal Component Analysis

Let's use `RunPCA` function of the Seurat. Pay attention to the `features` argument. How many features will PCA work on?

```{r pca, message=FALSE, warning=FALSE}
mgsc <- RunPCA(mgsc, npcs = 100, features = VariableFeatures(object = mgsc)) 
```

We can explore the constructed embeddings via  `mgsc@reductions`.

```{r reductions}
    slotNames(mgsc@reductions[["pca"]])
    dim(mgsc@reductions[["pca"]]@cell.embeddings)
    mgsc@reductions[["pca"]]@cell.embeddings[1:5,1:5]
    # OR
    # Embeddings(mgsc, reduction = "pca")[1:5, 1:5]
    
```

**Q5: Do you remember what kind of information feature.loadings store? ** 

<details>
<summary>**Answer**</summary>
`r colorize("feature.loadings: stores the weight for each feature along each dimension of the embedding.", "red")` 
</details>

```{r pcaviz, message=FALSE, warning=FALSE}

  # (1) we can visualize top genes associated with pca embeddings
  VizDimLoadings(mgsc, dims = 1:2, reduction = "pca")
  VizDimLoadings(mgsc, dims = 3:4, reduction = "pca")
```


### Determine the ‘dimensionality’ of the dataset

**Q6: Name that one method that we used to determine the dimensionality. Let's try to find out how much of the variance is explained by the PCs. ** (` Hint:` check `slotNames(mgsc@reductions[["pca"]])` which feature might help you with this. )

<details>
<summary>**Answer**</summary>
```{r elbow }
percent <- 100*mgsc@reductions[["pca"]]@stdev^2/sum(mgsc@reductions[["pca"]]@stdev^2)

perc_data <- data.frame(percent=percent, PC=1:length(percent))
ggplot(perc_data, aes(x=PC, y=percent)) + 
  geom_line(aes(y= percent), linetype=2) + 
  scale_x_continuous(breaks=seq(from=0, to=80, by= 10)) +
  geom_text(aes(label=round(percent, 2)), size=2, vjust=-.5) + 
  ylim(0, 80)

# OR you can directly use Seurat function 
ElbowPlot(mgsc,  ndims = 80)
```
</details>

<!--
Another alternative is to use `JackStraw` function to decide the number of dimensions. 

The `JackStrawPlot` function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). **Significant PCs** will show a strong enrichment of features with low p-values (solid curve above the dashed line). 


```{r jackstraw}
    mgsc <- JackStraw(mgsc, dims=50, num.replicate = 100) 
    mgsc <- ScoreJackStraw(mgsc, dims = 1:50)
    JackStrawPlot(mgsc, dims = 1:20)
```

This technique, however, might be time-consuming - especially for larger datasets. Elbow plot is a common practice for its speed. 

--> 

**Q7: Now it's time to visualize our data!  Generate the following output: three 2D plots with the first six PCs and print them side by side. i.e. PC1-PC2, PC3-PC4, PC5-PC6. ** (`Hint:` Utilize `DimPlot` and `plot_grid` functions.) 

<details>
<summary>**Answer**</summary>
```{r pcaprint }
  # (2) 
  plot_grid(ncol = 3,
    DimPlot(mgsc, reduction = "pca", dims = 1:2) + theme(legend.position="none"),
    DimPlot(mgsc, reduction = "pca", dims = 3:4) + theme(legend.position="none"),
    DimPlot(mgsc, reduction = "pca", dims = 5:6) + theme(legend.position="none") )
```
</details>


## Non-linear dimensional reduction (t-SNE/UMAP)

### t-SNE

Seurat comprises `RunTSNE` which uses `Rtsne` library (which we previously worked with in the lecture). 

```{r runtsne, message=FALSE, warning=FALSE}
# we are using the PCA embeddings as our input
mgsc <- RunTSNE(mgsc, reduction="pca", dims = 1:30,  nthreads = 4, max_iter = 2000)
mgsc@reductions
```

```{r tsneplot, message=FALSE, warning=FALSE}
DimPlot(mgsc, reduction = "tsne")

```

**Q8:Do you remember one of the important hyper-parameters of t-SNE?  Print t-SNE plots using `DimPlot` and explore  how structure of the printed data changes with this parameter. Keep dimension as $30$. ** (` Hint:` `reduction.name` will help you save your embeddings that are produced by different configurations. )


```{r tsneperp, echo=FALSE, message=FALSE,  warning=FALSE}

mgsc <- RunTSNE(mgsc, dims = 1:30, perplexity=5, reduction.name = "tsne_p5", nthreads = 4, max_iter = 2000)
mgsc <- RunTSNE(mgsc, dims = 1:30, perplexity=15, reduction.name = "tsne_p15", nthreads = 4, max_iter = 2000)
mgsc <- RunTSNE(mgsc, dims = 1:30, perplexity=30, reduction.name = "tsne_p30_default", nthreads = 4, max_iter = 2000)
mgsc <- RunTSNE(mgsc, dims = 1:30, perplexity=50, reduction.name = "tsne_p50", nthreads = 4, max_iter = 2000)

plot_grid(nrow=2, ncol = 2,
  DimPlot(mgsc, reduction = "tsne_p5")+ ggtitle(label ="p=5") + theme(legend.position="none"),
  DimPlot(mgsc, reduction = "tsne_p15")+ ggtitle(label ="p=15") + theme(legend.position="none"),
  DimPlot(mgsc, reduction = "tsne_p30_default")+ ggtitle(label ="p=30, default") + theme(legend.position="none"),
  DimPlot(mgsc, reduction = "tsne_p50")+ ggtitle(label ="p=50") + theme(legend.position="none")
)

```


### UMAP

```{r umap, , message=FALSE, warning=FALSE}
mgsc <- RunUMAP(mgsc, dims = 1:30)
DimPlot(mgsc, reduction = "umap")
```

**Homework: How about the important parameters for UMAP? Print following UMAP plots using `DimPlot` and explore how structure of the visualized data changes with these two parameters. Choose dimensions between $1:30$. ** 

```{r umapneighbor, echo=FALSE, message=FALSE, warning=FALSE}

mgsc <- RunUMAP(mgsc, reduction.name = "UMAP_n5", n.neighbors=5, dims = 1:30)
mgsc <- RunUMAP(mgsc, reduction.name = "UMAP_n20", n.neighbors=20, dims = 1:30)
mgsc <- RunUMAP(mgsc, reduction.name = "UMAP_n30_default",  dims = 1:30) 
mgsc <- RunUMAP(mgsc, reduction.name = "UMAP_n40", n.neighbors=40,  dims = 1:30) 

plot_grid(nrow=2, ncol = 2,
  DimPlot(mgsc, reduction = "UMAP_n5")+ ggtitle(label ="n=5") + theme(legend.position="none"),
  DimPlot(mgsc, reduction = "UMAP_n20")+ ggtitle(label ="n=20") + theme(legend.position="none"),
  DimPlot(mgsc, reduction = "UMAP_n30_default")+ ggtitle(label ="n=30, default") + theme(legend.position="none"),
  DimPlot(mgsc, reduction = "UMAP_n40")+ ggtitle(label ="n=40") + theme(legend.position="none")
)

```


**Homework: Plot the projections of the dataset with three of the dimensionality reduction techniques printed side by side and explore `AugmentPlot`. **

```{r compare, echo=FALSE}
    options(repr.plot.width=12, repr.plot.height=4)
    
    p1 <- DimPlot(mgsc, reduction = "tsne", pt.size = 0.1) + ggtitle(label = "t-SNE") 
    p2 <- DimPlot(mgsc, reduction = "umap", pt.size = 0.1) + ggtitle(label = "UMAP")
    p3 <- DimPlot(mgsc, reduction = "pca", pt.size = 0.1) + ggtitle(label = "PCA")
    
    p1 <- AugmentPlot(plot = p1 )
    p2 <- AugmentPlot(plot = p2 )
    p3 <- AugmentPlot(plot = p3 )
    (p1 + p2 + p3) & NoLegend()
```




# Cluster the cells

For our subset of time point E6.5, we have the cell type annotation information. Although this is not usually the case in many computational problems, i.e. we start without knowing (or having a very vague idea of) how many clusters we will end up. For now let's take advantage of the known annotations. 

**Q9: How many $real$ classes we have? Reproduce the following plots colored by cell type.** ( `Hint:` You can make use `group.by` argument in `DimPlot` to extract stored cluster IDs.)

```{r realclasses, echo=FALSE, message=FALSE, warning=FALSE}
    # length(unique(mgsc@meta.data$celltype))
    
    options(repr.plot.width=14, repr.plot.height=6)
    p1 <-DimPlot(mgsc, reduction = "umap", group.by = "celltype")+ggtitle("UMAP cell types") 
    p2 <-DimPlot(mgsc, reduction = "tsne", group.by = "celltype")+ggtitle("t-SNE cell type") + theme(legend.position="none")
    p1 + p2
    
    unique(mgsc@meta.data$celltype)
```


## K-means 

Seurat does not support K-means? What are we going to do now!? :)

**Q10: What do we need to run $K-means$? Try to reproduce the following plot where true classes are compared to k-means clusters ($k=13$).** ( `Hint:` Make use of `Embeddings` function or utilize `reductions` we previously used in the previous examples. You can save cluster IDs to `mgsc@meta.data` field. )


```{r kmeans, echo=FALSE}
pc30 <- Embeddings(object = mgsc, reduction = "pca") [, 1:30]
mgsc_kmeans <-kmeans(pc30,  iter.max = 100, centers=13)

dim(pc30)
# add clustering results to 'object@meta.data$kmeans.clusters'
mgsc@meta.data$kmeans.clusters <- as.factor(mgsc_kmeans$cluster)

p1 <- DimPlot(mgsc, reduction = "tsne", group.by = "celltype")+ggtitle("tSNE cell types") + theme(legend.position="none")
p2 <- DimPlot(mgsc, reduction = "tsne", group.by = "kmeans.clusters", label=TRUE )+ggtitle("tSNE k-means clusters") + theme(legend.position="none")

p3 <- DimPlot(mgsc, reduction = "umap", group.by = "celltype")+ggtitle("UMAP cell types") + theme(legend.position="none")
p4 <- DimPlot(mgsc, reduction = "umap", group.by = "kmeans.clusters", label=TRUE)+ggtitle("UMAP k-means clusters") + theme(legend.position="none")

plot_grid(nrow=2, ncol = 2, 
          p1, p2, 
          p3, p4)

```

**Homework: Explore how clusters would have changed if you used t-SNE embeddings instead of PCA as an input to k-means algorithm.**

```{r kmeanstsne, echo=FALSE}
tsne2 <- Embeddings(object = mgsc, reduction = "tsne") 
mgsc_kmeans <-kmeans(tsne2,  iter.max = 100, centers=13)

dim(tsne2)
# add clustering results to 'object@meta.data$kmeans.clusters'
mgsc@meta.data$kmeans.clusters.tsne <- as.factor(mgsc_kmeans$cluster)


p1 <- DimPlot(mgsc, reduction = "pca", group.by = "kmeans.clusters.tsne", label=TRUE)
p2 <- DimPlot(mgsc, reduction = "tsne", group.by = "kmeans.clusters.tsne", label=TRUE)
p3 <- DimPlot(mgsc, reduction = "umap", group.by = "kmeans.clusters.tsne", label=TRUE)
p1 + p2 + p3

```


### How many clusters should we choose?

We already know the number of cell-types in our data. But this is not always the case. How would we decide the number of clusters if we didn't know the number of classes?  Which $k$-means feature did we look at the to decide number of clusters? 

**Homework: Print the following plot using the same embeddings from the previous example. **  (`Hint:` Lecture notes:))

```{r clusternum, echo=FALSE}
    wss_ <- list() 
    clus <- seq(4, 40, by=4)
    
    for (k in 1:length(clus)) {
        m_kmeans <- kmeans(pc30, iter.max = 100, centers=clus[k])
        wss_ <- append(wss_, m_kmeans$tot.withinss)
        }
    
    wss_data <- data.frame(wss=unlist(wss_), k=clus)
    ggplot(wss_data, aes(x=k, y=wss)) + 
      geom_line(aes(y= wss), linetype=2) + 
      scale_x_continuous(breaks=seq(from=0, to=64, by= 4)) +
      ylim(100000, 250000)
```

Can we decide the number of clusters looking at this plot?

## Hierarchical clustering

Seurat does not support hierarchical clustering too? But we can do it!! 

**Q11:Which elements do we need to perform hierarchical clustering? Try to reproduce the following dendogram. Use `Euclidian` as distance metric and `ward.D2` as linkage method. **  


```{r hclust, echo=FALSE}
mgsc_dist <- dist( mgsc@reductions[["pca"]]@cell.embeddings[,1:30], method="euclidean")
mgsc_hce <- hclust(mgsc_dist, method="ward.D2")

# Plot the obtained dendrogram - this might take time!!
plot(mgsc_hce, cex = 0.6, hang = -1, main='mgsc cluster dendogram')

```

Looking at the dendogram, we can investigate different numbers of clusters using `cutree` function that cuts the hierarchical clustering tree  into given number of clusters. 

**Q12: How about we print different clustering outcomes with $k=4, 8, 16$ using t-SNE? **


```{r hclustplot, echo=FALSE}
#euclidean distance
mgsc@meta.data$hce_4 <- cutree(mgsc_hce,k = 4)
mgsc@meta.data$hce_8 <- cutree(mgsc_hce,k = 8)
mgsc@meta.data$hce_16 <- cutree(mgsc_hce,k = 16)

options(repr.plot.width=21, repr.plot.height=5)

plot_grid(ncol = 3,
  DimPlot(mgsc, reduction = "tsne", group.by = "hce_4", label=TRUE)+ggtitle("k=4"),
  DimPlot(mgsc, reduction = "tsne", group.by = "hce_8", label=TRUE)+ggtitle("k=8"),
  DimPlot(mgsc, reduction = "tsne", group.by = "hce_16", label=TRUE)+ggtitle("k=16"))
```


## Graph-based clustering

Seuratv3 adopts a graph-based clustering methodology much similar to PhenoGraph approach we discussed earlier.

`FindNeighbors ` function performs the following steps: 
* (1) build a kNN graph using Euclidean distance in PCA space
* (2) update the edge weights based on the shared neighbors (Jaccard index) to construct a Shared Nearest Neighbor (SNN) graph. 

Once the graph is constucted, we can use  `FindClusters ` function  to apply a community detection algorithm to identify subgroups. Keep in mind that Seurat uses the Louvain algorithm as a default for this step.


```{r snn, message=FALSE, warning=FALSE}
mgsc <- FindNeighbors(mgsc,  k.param = 20, dims = 1:30, reduction = "pca")
mgsc <- FindClusters(mgsc, resolution = 0.5)


# Look at cluster IDs of the first 20 cells
head(Idents(mgsc), 20)
```

`FindClusters ` has a resolution parameter that sets the `granularity` of the downstream clustering, with increased values leading to a greater number of clusters.  

The Seurat authors suggest that  `granularity values` between 0.4-1.2 usually return good results for sc datasets of around 3K cells. For larger datasets, optimal resolution often increases for larger datasets. 


**Homework:** Explore how  different values of $granularity$  affect the clustering results and reproduce the following plot. (Hint: Use `Idents` function of Seurat to save cluster ids. )


```{r snnplot, message=FALSE,  warning=FALSE, echo=FALSE}
  
  mgsc[["org.ident"]] <- Idents(object = mgsc) #0.5
  mgsc <- FindClusters(mgsc, resolution = 0.4)
  mgsc[["ident.04"]] <- Idents(object = mgsc)
  mgsc <- FindClusters(mgsc, resolution = 0.8)
  mgsc[["ident.08"]] <- Idents(object = mgsc)
  mgsc <- FindClusters(mgsc, resolution = 1.2)
  mgsc[["ident.12"]] <- Idents(object = mgsc)

  
  plot_grid(nrow=2, ncol = 2,
  DimPlot(mgsc, reduction = "umap", group.by = "ident.04", label=TRUE)+ggtitle("SNN res=0.4"),
  DimPlot(mgsc, reduction = "umap", group.by = "org.ident", label=TRUE)+ggtitle("SNN res=0.5, default"),
  DimPlot(mgsc, reduction = "umap", group.by = "ident.08", label=TRUE)+ggtitle("SNN res=0.8"),
  DimPlot(mgsc, reduction = "umap", group.by = "ident.12", label=TRUE)+ggtitle("SNN res=1.2")
  
  )
  
```







