---
title: "Seurat Basic Tutorial"
output:
  html_document:
    df_print: paged
---

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
```


Data can be download here: 
https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz
Download and extract it in a folder of your choice and replace the directory in the following chunk. 

We read the data from the 10x output. There is already a function that can handle this type of directories.
You can go to the folder and check the data that is there stored by doing "ls" in the terminal.


```{r}
pbmc.data <- Read10X(data.dir = "/home/r511a/projects/Gonzalo/scCourse2021/filtered_gene_bc_matrices/hg19")
```

We can create a Seurat object by just calling the CreateSeuratObject() function. It will receive as parameters the "counts" matrix, the name of the "project" and then we can provide thresholds for filtering the data based on some properties. 

min.cells tells us which is the minimum number of cells in which a feature (gene in this case) needs to be present in order to keep that feature.

min.features is the minimum number of features (genes in this case) that need to be detected for a cell to be kept in the data. 

```{r}
min.cells<-3
min.features<-200

pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = min.cells, min.features = min.features)
```

There are many other ways of filtering low quality cells. One very common practice is to check the percentage of mithocondrial genes per cell. Cells with high numbers of mithocondrial genes are considered to be under stress and hence are low quality cells. In order to filter out cells with high mithocondrial genes contents we quantify the occurrence of certain genes in the data. In case we are workin with gene symbols, mithocondrial genes are the ones for which their names start by "MT-".

```{r}
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
```

It might be possible that instead of gene symbols we get our genes with Ensemble GeneIDs. In that case we will need to map to gene symbols first. 

This can be done by:
library(biomaRt)
ensembl <- useMart("ensembl", dataset="mmusculus_gene_ensembl")
annot<-getBM(c("ensembl_gene_id", "mgi_symbol", "chromosome_name", "strand", "start_position", "end_position","gene_biotype"), mart=ensembl)

Once we get the mapping tables we can subselect the genes and continue with the following steps. 

After we calculated the mithocondrial percentage it will be stored as a feature in the Seurat object to which we can refer later. 

We can check the distribution of different features accross cells visually by doing:

```{r}
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


When analysing the violing plots we want to filter out those cells that deviate from the majority of cells in the distribution of a given feature. Thresholds for these are dataset dependant. 

In our case we will filter as follows by using the function subset(). This function subselects a certain group of cells based on boolean conditions that we set according to the different thresholds in the features that we have calculated across cells. 

```{r}
minRNA<-200
maxRNA<-2500
maxMTpercent<-5

pbmc <- subset(pbmc, subset = nFeature_RNA > minRNA & nFeature_RNA < maxRNA & percent.mt < maxMTpercent)
```

We can run the violing plots function again to see how now the outlier cells are not present in the data anymore. 
```{r}
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

After removing low quality cells the next step is to normalize the data:
By default, Seurat applies a global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result.

Remember why do we log transform?

```{r}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
```

More recent versions of Seurat have implemented a new way to normalize the date calles scTransform that replaces the log normalization. In that case instead of using the previous step, you can apply something like this:

```{r}
#pbmc <- SCTransform(pbmc, vars.to.regress = "percent.mt", verbose = FALSE)
```

We can now find those genes that vary the most accross the dataset.  Focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

```{r}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```

The last preprocessing step before reducing the dimensionality of the data is to Scale the data. Scaling the data means to transform the gene expression distributions to become z-distributions which will have a mean=0 and standard deviation = 1

```{r}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

Next Step we calculate Principal Components for our data. We can use all genes or a subset of them. In this case we use the variable genes that we detected before.

```{r}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
```

```{r}
DimPlot(pbmc, reduction = "pca")
```

We can visually inspect those genes that correlate the most with the PCs both positively and negatively.

```{r}
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```

How many PCA dimensions do we need to keep for downstream analysis?
An elbow plot helps us to decide on this. We can order principal components according to their eigen values or standard deviation. We want to keep those PCs that explain an important proportion of variability in the data. 

```{r}
ElbowPlot(pbmc)
```

Next we proceed to cluster the data. 

```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```

Based on our PCA analysis we would keep 10 components. The problem with that is that we can't visualize the PCA data from this 10 dimensions in an easy way. For this, we can summarize the PCA information with extra layers of data transformation that can apply different mathematical processing to the information that is captured in the 10 PCs and collapse it into 2/3 dimensions that we can easily visualize. 

```{r}
pbmc <- RunTSNE(pbmc, dims = 1:10)
```


```{r}
pbmc <- RunUMAP(pbmc, dims = 1:10)
```

We plot the data for the two embeddings and we show the clusters as they were detected on the PCA data. 

```{r}
plot1 <- DimPlot(pbmc, reduction = "tsne")
plot2 <-DimPlot(pbmc, reduction = "umap")
plot1 + plot2
```

```{r}
saveRDS(pbmc, file = "/home/r511a/projects/Gonzalo/scCourse2021/pbmc_tutorial_1stpart_output.rds")
```

We can check for markers for any of the identified clusters in our object
```{r}
cluster3.markers <- FindMarkers(pbmc, ident.1 = 3, min.pct = 0.25)
head(cluster3.markers, n = 5)
```

We can plot the genes we identify in our previous analysis across all clusters in our data.
```{r}
VlnPlot(pbmc, features = rownames(cluster3.markers)[1:2])
```

We can also check the expression distribution of the genes in the low dimensional embeddings.
```{r}
FeaturePlot(pbmc, features = rownames(cluster3.markers)[1:2])
```
We can find the gene markers for all clusters
```{r}
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
```

We can plot the top 10 markers per cluster

```{r}
top10 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```

Based on the markers we found we can rename our clusters to meaningful names
```{r}
new.cluster.ids <- c("Naive CD4 T", "Memory CD4 T", "CD14+ Mono", "B", "CD8 T", "FCGR3A+ Mono", 
    "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

```{r}
saveRDS(pbmc, file = "/home/r511a/projects/Gonzalo/scCourse2021/pbmc_tutorial_2ndpart_output.rds")
```
