---
title: "Multi-omics Factor Analysis applied to the gastrulation scNMT-seq data set"
author:
  name: "Ricard Argelaguet"
  affiliation: "European Bioinformatics Institute, Cambridge, UK"
  email: "ricard@ebi.ac.uk"
date: "`r Sys.Date()`"

output:
  BiocStyle::html_document:
    toc: true
---

# Description

This tutorial demonstrates how to use MOFA for the integration of single-cell multi-omics data.  
We consider a dataset where scNMT-seq was used to simultaneously profile RNA expression, DNA methylation and chromatin accessibility in 1,828 cells at multiple stages of mouse development. MOFA provides a method for delineating coordinated variation between the transcriptome and the epigenome.

As input to the model we quantified DNA methylation and chromatin accessibility values over two different sets of regulatory elements: gene promoters and enhancer elements. RNA expression was quantified over protein-coding genes. After data processing, separate views were defined for the RNA expression and for each combination of genomic context and epigenetic readout.

The data set we use here is a simplified version of the original data set (only E7.5 cells) published in (Nature)[https://www.nature.com/articles/s41586-019-1825-8]. The full data set can be downloaded from [this FTP](ftp://ftp.ebi.ac.uk/pub/databases/scnmt_gastrulation).


```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, fig.align="center")
```

# Load libraries

Load dependencies. Make sure that MOFA2 is imported last, to avoid collisions with functions from other packages
```{r, echo=FALSE}
library(data.table)  # fast manipulation of data.frames
library(purrr)       # pipes to make the code more readable
library(ggplot2)
library(MOFA2)
```

Define cell type colors for the visualisations
```{r echo = FALSE}
colors <- c(
  # "Epiblast" = "grey70",
  "Mesoderm" = "#CD3278",
  # "Primitive Streak" = "sandybrown",
  "Endoderm" = "#43CD80",
  "Ectoderm" = "steelblue"
  # "ExE Endoderm" = "#E066FF"
)
```

In this vignette we skip all the data processing details as well as the model training part. We focus on the downstream characterisation of the MOFA model. 

# Load pre-computed MOFA model

Note that for this tutorial I selected the MOFA Factors that explained at least 1% of variation in the RNA expression.


```{r}
# load(url("ftp://ftp.ebi.ac.uk/pub/databases/mofa/scnmt_gastrulation/gastrulation_scnmt_mofa.RData"))
MOFAobject <- readRDS("/Users/ricard/data/teaching_heidelberg/mofa/MOFAmodel.rds")
MOFAobject
```

```{r}
views_names(MOFAobject) <- c("Enhancers accessibility","Promoters accessibility", "Enhancers methylation","Promoters methylation", "RNA expression")
```

Explore the cell metadata:  
* **sample**: cell ID  
* **stage**: developmental stage.  
* **lineage**: cell type annotation (derived from mapping the cells to the [10x reference atlas](https://www.nature.com/articles/s41586-019-0933-9)).  
* **pass_rnaQC**: did the cell pass QC for RNA expression?.  
* **pass_metQC**: did the cell pass QC for DNA methylation? `NA` if the cell was only profiled for RNA.  
* **pass_accQC**: did the cell pass QC for chromatin accessibility? `NA` if the cell was only profiled for RNA.  
* **group**: ignore this column

```{r}
head(samples_metadata(MOFAobject))
```


# Overview of training data

The function `plot_data_overview` can be used to obtain an overview of the input data. 
It shows how many views (rows) and how many cells (columns) exist, what are their corresponding dimensionalities are. It also shows which views each cell is missing.
```{r fig.align="center"}
view.colors <- c(
  "RNA expression" = "#3CB54E",
  "Enhancers accessibility" = "#00BFC4",
  "Promoters accessibility" = "#00BFC4",
  "Enhancers methylation" = "#F37A71",
  "Promoters methylation" = "#F37A71"
)
view.colors = view.colors[views_names(MOFAobject)]

plot_data_overview(MOFAobject, colors = view.colors)
```

One can extract the input data from MOFA using the `get_data` function.

Visualise RNA expression data. Notice that it is already log-normalised, but it looks zero-inflated. This is a statistical challenge for *some* single-cell RNA-seq assays.
```{r}
rna <- get_data(MOFAobject, views="RNA expression")[[1]][[1]]
hist(rna)
```

Visualise DNA methylation and chromatin accessibility data. We use M-values instead of Beta-values. Notice that both data modalities are not well captured by a gaussian likelihood model, so we won't have a very good model fit. This data should be modelled with a binomial distribution: for every cell $i$ and region $j$ the total number of CpGs correspond to the total number of trials and the number of methylated CpGs to the number of successes. The binomial likelihood is not implemented in the MOFA framework and thus why we need to really on calculating M-values and then using a Gaussian distribution.
```{r}
met <- get_data(MOFAobject, views="met_Promoters")[[1]][[1]]
hist(met)
```

```{r}
acc <- get_data(MOFAobject, views="acc_Promoters")[[1]][[1]]
hist(acc)
```

# Overview of the MOFA model

## Correlation between factors

A good sanity check is to verify that the Factors are largely uncorrelated. In MOFA there are no orthogonality constraints such as in Principal Component Analysis, but if there is a lot of correlation between Factors this suggests a poor model fit.
```{r}
plot_factor_cor(MOFAobject)
```


## Variance decomposition analysis
**The most important insight that MOFA generates is the variance decomposition analysis**. This plot shows the percentage of variance explained by each factor in each data modality. It summarises the (latent) signal from a complex heterogeneous data set in a single figure.  

```{r}
plot_variance_explained(MOFAobject) +
  theme(
    axis.text.x = element_text(colour="black",size=rel(1.1), angle=25, hjust=1, vjust=1.05)
  )
```

What insights from the data can we learn just from inspecting this plot?  

- **Factor 1** and **Factor 2** capture strong sources of variability that drive significant amounts of variation in RNA expression and epigenetic status of enhancer elements, but not in promoter elements.
- **Factor 3** and  **Factor 5** capture a source of variation that is exclusive to the RNA expression. 
- **Factor 4** captures a strong source of variation that is shared between RNA expression and DNA methylation of enhancer elements.

**(Q) What do you notice about promoters? Could you propose some hypothesis?**

# Characterisation of Factors

There are a few systematic strategies to characterise the molecular signal that underlies each MOFA Factor and to relate them to existent sample covariates:

- **Association analysis between the sample metadata and the Factor values**: function `correlate_factors_with_covariates`
- **Inspection of factor values**: functions `plot_factor` (one factor at a time) and `plot_factors` (combinations of factors)
- **Inspection of the feature weights**: functions `plot_weights` (all weights), `plot_top_weights` (only the top weights)
- **Gene set enrichment analysis on the mRNA weights**: functions `run_enrichment`, followed by `plot_enrichment`.

## Characterisation of Factor 1

### Visualisation of Factor values 

 Let's plot Factor 1 values and color cells by lineage assignment.  Clearly, this factor captures the variation that is associated with the separation between Mesoderm (positive Factor values) and non-Mesoderm cells (negative Factor values).
 
```{r}
plot_factor(MOFAobject,
  factor = 1,
  color_by = "lineage", 
  add_violin = TRUE,
  dodge = TRUE
) + scale_fill_manual(values=colors)
```

**How do we interpret the factor values?**  
Each factor captures a different source of variability in the data. Mathematically, each Factor is defined by a linear combination of the input features. As the data is centered prior to running MOFA, each Factor ordinates cells along a one-dimensional axis that is centered at zero. Samples with different signs manifest opposite phenotypes along the inferred axis of variation, with higher absolute value indicating a stronger effect. Note that the interpretation of MOFA factors is analogous to the interpretation of the principal components in PCA.


### Visualisation of RNA weights

The weights provide a score for each gene on each factor. Genes with no association with the factor are expected to have values close to zero, whereas genes with strong association with the factor are expected to have large absolute values. The sign of the loading indicates the direction of the effect: a positive loading indicates that the feature is more active in the cells with positive factor values, and viceversa.  

Let's plot the distribution of weights for Factor 1.
```{r, warnings=FALSE, message=FALSE}
plot_weights(MOFAobject,
  view = "RNA expression",
  factor = 1,
  nfeatures = 10,     # Top number of features to highlight
  scale = T           # Scale weights from -1 to 1
)
```

If you are not interested in the full distribution, but just on the top weights, you can instead do:
```{r}
plot_top_weights(MOFAobject, 
  view = "RNA expression", 
  factor = 1, 
  nfeatures = 10,
  scale = T, 
  abs = T
)
```

We expect that genes with large positive weights For Factor 1 to be highlighy expressed in the Mesoderm cells. If we plot Factor 1 colouring cells by gene expresion of the top genes *with positive weight*:
```{r}
genes <- c("Phlda2","Mesp1")

for (i in genes) {
  plot_factor(MOFAobject,
    factor = 1,
    dot_size = 2.5,
    color_by = i
  ) %>% print
}
```


Similarly, we expect that genes with large negative weights For Factor 1 to be lowly expressed in the Mesoderm cells. If we plot Factor 1 colouring cells by gene expresion of the top genes *with negative weight*:
```{r}
genes <- c("Cldn6","Pim2")

for (i in genes) {
  plot_factor(MOFAobject,
    factor = 1,
    dot_size = 2.5,
    color_by = i
  ) %>% print
}
```

## Visualisation of RNA expression patterns

The weights are useful to identify which genes are driving each factors However, after inspecting the weights it is good practice to go back to the high-dimensional space and check if the variability that MOFA captures is real.  
For example, one could generate a heatmap plot of the RNA expression for the top genes, where samples are sorted by the corresponding factor values. This is the aim of the `plot_data_heatmap` function:

```{r}
plot_data_heatmap(MOFAobject, 
  view = "RNA expression", 
  factor = 1, 
  features = 25,
  annotation_samples = "lineage",
  # extra arguments passed to `pheatmap`,
  show_colnames = F, cluster_cols = F, 
  annotation_colors = list("lineage"=colors),
  annotation_legend = FALSE
)
```

An interesting option of `plot_data_heatmap` is to plot "denoised" observations. This is obtained by reconstructing the data using the matrix factorisation equation from MOFA: $$\mathbf{Y} = \mathbf{W}\mathbf{Z}$$. This essentially removes all the variation that is not captured by the model:

```{r}
plot_data_heatmap(MOFAobject, 
  view = "RNA expression", 
  factor = 1, 
  denoise = TRUE,
  features = 25,
  show_colnames = F, cluster_cols = F # extra arguments passed to `pheatmap`
)
```

## Visualisation of DNA methylation weights

As we have done with RNA, we can also visualise the distribution of weights for the epigenetic modalities. The problem about this is that the large majority of enhancers are not well annotated and we only have the genomic coordinates for them...

```{r}
plot_weights(MOFAobject,
  view = c("Enhancers methylation"),
  factor = 1,
  nfeatures = 5,
  scale = F
)
```

## Visualisation of DNA methylation patterns

Let's visualise the coordinated variation that MOFA captures using a heatmap. We will select the top features with largest loading and plot its DNA methylation levels, with sampels ordered according to Factor 1. In addition, here we are going to use the imputation capabilities of MOFA to obtain a cleaner representation.  

First, we need to impute missing values:

```{r}
MOFAobject <- impute(MOFAobject)
```

Generate Heatmap of the top 25 features with largest loading, before imputation (`impute=FALSE`):

```{r, out.width="120%"}
plot_data_heatmap(MOFAobject, 
  view = "Enhancers methylation", 
  factor = 1, 
  impute = FALSE,
  features = 25,
  show_colnames = F, cluster_cols = F, 
  annotation_colors = list("lineage"=colors),
  annotation_legend = FALSE
)
```

After imputation (`impute=TRUE`):

```{r}

```


## Characterisation of Factor 2

**(Q) Your task is to provide a characterisation for Factor 2**.  
Try a similar pipeline as for Factor 1 and answer the following questions:
- Which germ layer underlies Factor 2?
- Can you identify mRNA markers? Validate them using the [scRNA-seq reference atlas](https://marionilab.cruk.cam.ac.uk/MouseGastrulation2018)
- Generate a heatmap that displays the DNA methylation variation (try with and without imputation of missing values, see the function `impute`) 

## Plot combinations of Factors

```{r}
plot_factors(MOFAobject,
  factors = c(1,2), 
  color_by = "lineage",
  dot_size = 2,
  legend = F
) + scale_fill_manual(values=colors)
```

This scatter plot captures in 2D all the variation that is required to separate the three germ layers! It corresponds to [Figure 2b in the paper](https://www.nature.com/articles/s41586-019-1825-8).  

We can colour this plot by the molecular measurements, for example gene expression:
```{r}
# Ectoderm marker
plot_factors(MOFAobject,
  factors = c(1,2), 
  color_by = "Pim2", 
  dot_size = 2
)

# Mesoderm marker
plot_factors(MOFAobject,
  factors = c(1,2), 
  color_by = "Mesp1", 
  dot_size = 2
)

# Endoderm marker
plot_factors(MOFAobject,
  factors = c(1,2), 
  color_by = "Foxa2", 
  dot_size = 2
)
```


# Ad hoc analysis with the factors and the weights

## Are the changes in DNA methylation and chromatin accessibility correlated at the loci level?

With MOFA we have identified coordinated axis of variation between the three omics (Factor1 and Factor2). However, this does not necessarily imply that the *same* features are driving the signal in all of the omics. It could be that DNA methylation changes associated with endoderm commitment (Factor 2) are driven by enhancers A,B,C whereas chromatin accessibility changes are driven by enhancers D,E,F. 
To explore this, we will correlate the weights for both epigenetc modalities. This cannot be done internally within MOFA so we have to extract the weights manually do our own downstream analysis This is done with the function `get_weights`:

```{r}
# Fetch weights
w.met <- get_weights(MOFAobject, 
  factors = c(1,2), 
  views = "Enhancers methylation", 
  scale = TRUE,
  as.data.frame = T
) %>% as.data.table

w.acc <- get_weights(MOFAobject, 
  factors = c(1,2), 
  views = "Enhancers accessibility", 
  scale = TRUE,
  as.data.frame = T
) %>% as.data.table

# Remove the met_ and acc_ prefix from the feature names
w.met[,feature:=substr(feature,5,nchar(as.character(feature)))]
w.acc[,feature:=substr(feature,5,nchar(as.character(feature)))]

# Merge weights
w.dt <- merge(
  w.met[,c("feature","factor","value")], 
  w.acc[,c("feature","factor","value")], 
  by=c("feature","factor")
)
```

Notice how the DNA methylation and chromatin accessibility weighs are highly correlated, suggesting that the variability in both epigenetic layers is highly coupled not just globally but also at the individual loci level.
```{r}
ggplot(w.dt, aes(x=value.x, y=value.y)) +
  geom_point() + 
  stat_smooth(method="lm") +
  theme_classic() +
  facet_wrap(~factor, scales="free") +
  labs(x="DNA methylation weight", y="Chromatin accessibility weight")
```

# sessionInfo()

```{r}
sessionInfo()
```

